### PR TITLE
ENH: Support scalar condition in `Series.where` and `DataFrame.where`

### LINKED ISSUE #53903: ENH: Support scalar condition in `Series.where` and `DataFrame.where`
### Feature Type

- [x] Changing existing functionality in pandas

### Problem Description

For compatibility reasons with [`numpy.where`](https://numpy.org/doc/stable/reference/generated/numpy.where.html) it would be nice to allow scalar booleans as a condition to `Series.where` and `DataFrame.where`. Currently

```python
import pandas as pd
s = pd.Series([1,2])
s.where(True)
```

raises `ValueError: Array conditional must be same shape as self`

### Feature Description

Allow scalar boolean as input to 

- `Series.where`
- `Series.mask`
- `DataFrame.where`
- `DataFrame.mask`

### Alternative Solutions

idk.

### Additional Context

Regarding documentation, I am not sure anything needs to be updated since both `Series.where` and `DataFrame.where` specify `array-like` as a valid input type, and scalars are considered `array-like`.

### PR DESCRIPTION
**Pull request by randolf-scholz**
- [x] closes #53903
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
    - added/updated tests in `series/indexing/test_where.py`, `series/indexing/test_where.py`, `series/indexing/test_mask.py`
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
  - Regarding documentation, I am not sure anything needs to be updated since both Series.where and DataFrame.where specify array-like as a valid input type, and scalars are considered array-like.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.1.0.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -10102,8 +10102,15 @@ def _where(
         else:
             if not hasattr(cond, "shape"):
                 cond = np.asanyarray(cond)
-            if cond.shape != self.shape:
-                raise ValueError("Array conditional must be same shape as self")
+            if cond.shape == ():
+                # Note: DataFrame(True, index=[1,2,3], columns=["a", "b", "c"]) works
+                # but DataFrame(np.array(True), index=[1,2,3], columns=["a", "b", "c"])
+                # does not hence we need to unpack scalar
+                cond = cond.item()
```
Comment (jbrockmendel) [inline pandas/core/generic.py:10109]: use lib.item_from_zerodim here?

Comment (randolf-scholz) [reply pandas/core/generic.py]: Done.

**Comment by randolf-scholz** [comment]
@jbrockmendel Any other objections?

**Comment by jbrockmendel** [comment]
> Any other objections?

Not about the implementation, I'm just not sure I want to support this.  In general we're trying to move towards being stricter in signatures.  But not strong enough an opinion to cause a stink here.

**Comment by mroeschke** [comment]
Sounds like this feature needs more discussion and buy in in the original issue so going to close for now awaiting that decision. Can reopen once that has been done

**Pull request closed without merging**